apiserver/pkg/storage/calico/blockAffinity_storage.go

```diff
@@ -0,0 +1,129 @@
+	"k8s.io/apiserver/pkg/storage/storagebackend/factory"
+
+	"github.com/projectcalico/calico/libcalico-go/lib/clientv3"
+	"github.com/projectcalico/calico/libcalico-go/lib/options"
+	"github.com/projectcalico/calico/libcalico-go/lib/watch"
+
+	libapi "github.com/projectcalico/calico/libcalico-go/lib/apis/v3"
+
+	aapi "github.com/projectcalico/api/pkg/apis/projectcalico/v3"
+)
+
+// NewBlockAffinityStorage creates a new libcalico-based storage.Interface implementation for BlockAffinity
+func NewBlockAffinityStorage(opts Options) (registry.DryRunnableStorage, factory.DestroyFunc) {
+	c := CreateClientFromConfig()
+	createFn := func(ctx context.Context, c clientv3.Interface, obj resourceObject, opts clientOpts) (resourceObject, error) {
+		return nil, fmt.Errorf("Unable to create block affinity. Block affinity resources are read-only.")
+	}
+	updateFn := func(ctx context.Context, c clientv3.Interface, obj resourceObject, opts clientOpts) (resourceObject, error) {
+		return nil, fmt.Errorf("Unable to update block affinity. Block affinity resources are read-only.")
+	}
+	getFn := func(ctx context.Context, c clientv3.Interface, ns string, name string, opts clientOpts) (resourceObject, error) {
+		ogo := opts.(options.GetOptions)
+		return c.BlockAffinities().Get(ctx, name, ogo)
+	}
+	deleteFn := func(ctx context.Context, c clientv3.Interface, ns string, name string, opts clientOpts) (resourceObject, error) {
+		return nil, fmt.Errorf("Unable to delete block affinity. Block affinity resources are read-only.")
+	}
+	listFn := func(ctx context.Context, c clientv3.Interface, opts clientOpts) (resourceListObject, error) {
+		olo := opts.(options.ListOptions)
+		return c.BlockAffinities().List(ctx, olo)
+	}
+	watchFn := func(ctx context.Context, c clientv3.Interface, opts clientOpts) (watch.Interface, error) {
+		olo := opts.(options.ListOptions)
+		return c.BlockAffinities().Watch(ctx, olo)
+	}
+	dryRunnableStorage := registry.DryRunnableStorage{Storage: &resourceStore{
+		client:            c,
+		codec:             opts.RESTOptions.StorageConfig.Codec,
+		versioner:         etcd.APIObjectVersioner{},
+		aapiType:          reflect.TypeOf(aapi.BlockAffinity{}),
+		aapiListType:      reflect.TypeOf(aapi.BlockAffinityList{}),
+		libCalicoType:     reflect.TypeOf(libapi.BlockAffinity{}),
+		libCalicoListType: reflect.TypeOf(libapi.BlockAffinityList{}),
+		isNamespaced:      false,
+		create:            createFn,
+		update:            updateFn,
+		get:               getFn,
+		delete:            deleteFn,
+		list:              listFn,
+		watch:             watchFn,
+		resourceName:      "BlockAffinity",
+		converter:         BlockAffinityConverter{},
+	}, Codec: opts.RESTOptions.StorageConfig.Codec}
+	return dryRunnableStorage, func() {}
+}
+
+type BlockAffinityConverter struct {
+}
+
+func (gc BlockAffinityConverter) convertToLibcalico(aapiObj runtime.Object) resourceObject {
+	aapiBlockAffinity := aapiObj.(*aapi.BlockAffinity)
```

Comment:
Strictly speaking, this function should never be called, right? Since we only ever do read operations?

If so, we should probably just stub it out and log an error if it is called.